common/rate-limiter/src/rate_limit.rs

```diff
@@ -0,0 +1,446 @@
+/// tokens added to it will "spill" out of the `bucket`.  The `buckets` are filled at an `interval`
+/// with a given `fill rate`.
+///
+/// ## Interval
+/// The `interval` at which we refill *all* of the `buckets` in the token bucket filter. Configured
+/// across the whole token bucket filter.
+///
+/// ## Fill Rate
+/// The rate at which we fill a `bucket` with tokens. Configured per bucket.
+///
+/// ## Bucket Size
+/// Maximum size of a bucket.  A bucket saturates at this size.  Configured per bucket.
+///
+/// # Features
+/// ## Keys
+/// The token bucket takes any key as long as it's hashable.  This should allow it to apply to
+/// many applications that need rate limiters.
+///
+/// ## Bucket sizes and Rates
+/// ### Defaults
+/// There are defaults for bucket size and fill rate, which will apply to unknown keys.
+///
+/// ### Refill Interval
+/// Buckets are refilled automatically at an interval.  To do this synchronously, it calculates the
+/// number of intervals that have passed.  This is done synchronously and in the future may be done
+/// asynchronously.
+///
+pub struct TokenBucketRateLimiter<Key: Eq + Hash + Clone + Debug> {
+    buckets: RwLock<HashMap<Key, Arc<Mutex<Bucket>>>>,
+    default_bucket_size: NonZeroUsize,
+    default_fill_rate: NonZeroUsize,
+    open: bool,
+}
+
+impl<Key: Eq + Hash + Clone + Debug> TokenBucketRateLimiter<Key> {
+    pub fn new(default_bucket_size: NonZeroUsize, default_fill_rate: NonZeroUsize) -> Self {
+        assert!(
+            default_bucket_size >= default_fill_rate,
+            "Bucket size must be greater than or equal to fill rate"
+        );
+        Self {
+            buckets: RwLock::new(HashMap::new()),
+            default_bucket_size,
+            default_fill_rate,
+            open: false,
+        }
+    }
+
+    /// Used for testing and to not have a rate limiter
+    pub fn open() -> Self {
+        let max = NonZeroUsize::new(std::usize::MAX).unwrap();
+        Self {
+            buckets: RwLock::new(HashMap::new()),
+            default_bucket_size: max,
+            default_fill_rate: max,
+            open: true,
+        }
+    }
+
+    /// Retrieve bucket, or create a new one
+    pub fn bucket(&self, key: Key) -> Arc<Mutex<Bucket>> {
+        if !self.open {
+            self.bucket_inner(key, |size, rate| {
+                Arc::new(Mutex::new(Bucket::new(size, rate)))
+            })
+        } else {
+            self.bucket_inner(key, |_, _| Arc::new(Mutex::new(Bucket::open())))
+        }
+    }
+
+    fn bucket_inner<F: FnOnce(NonZeroUsize, NonZeroUsize) -> Arc<Mutex<Bucket>>>(
+        &self,
+        key: Key,
+        bucket_create: F,
+    ) -> Arc<Mutex<Bucket>> {
+        // Attempt to do a weaker read lock first, followed by a write lock if it's missing
+        // For the common (read) case, there should be higher throughput
+        let maybe_value = { self.buckets.read().get(&key).cloned() };
+
+        if let Some(value) = maybe_value {
```

Comment:
why not merge these two lines?

Reply:
I separated it to ensure the read lock was let go immediately, if a write lock needed to take place after.

Let me circle back and see if I really needed it like this.

Reply:
Yep, basically without it being in a separate line, the lifetime of the read lock holds through the lower line write lock, which causes it to lock up on the same thread by never letting go of the read lock before trying to request a write lock.